### PR TITLE
 Add E2E tests for advanced cluster API

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

This PR introduces some new E2E tests for advanced clusters.

I've also fixed a few issues that appeared as a result of the tests.

* In certain configurations, advanced clusters will show up when using the regular cluster API (one's that are not spread across regions).
* E2E tests now have different logic to verify advanced clusters using the atlas api.
* Use reconciler has been updated to handle advanced cluster only deployments.

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

### CODE REVIEW

File: pkg/controller/atlasdatabaseuser/databaseuser.go [hunk 1]
```diff
@@ -149,28 +151,27 @@ func performUpdateInAtlas(ctx *workflow.Context, k8sClient client.Client, projec
 func validateScopes(ctx *workflow.Context, projectID string, user mdbv1.AtlasDatabaseUser) error {
 	for _, s := range user.GetScopes(mdbv1.ClusterScopeType) {
 		var apiError *mongodbatlas.ErrorResponse
-		_, _, err := ctx.Client.Clusters.Get(context.Background(), projectID, s)
-		if errors.As(err, &apiError) && apiError.ErrorCode == atlas.ClusterNotFound {
+		_, _, regularErr := ctx.Client.Clusters.Get(context.Background(), projectID, s)
+		_, _, advancedErr := ctx.Client.AdvancedClusters.Get(context.Background(), projectID, s)
```
Comment (chatton) [inline pkg/controller/atlasdatabaseuser/databaseuser.go:155]: the cluster could be either regular or advanced, we need to check for both

File: pkg/controller/atlasdatabaseuser/databaseuser.go [hunk 2]
```diff
@@ -149,28 +151,27 @@ func performUpdateInAtlas(ctx *workflow.Context, k8sClient client.Client, projec
...
+		if errors.As(regularErr, &apiError) && apiError.ErrorCode == atlas.ClusterNotFound && errors.As(advancedErr, &apiError) && apiError.ErrorCode == atlas.ClusterNotFound {
 			return fmt.Errorf(`"scopes" field references cluster named "%s" but such cluster doesn't exist in Atlas'`, s)
 		}
 	}
 	return nil
 }
 
 func checkClustersHaveReachedGoalState(ctx *workflow.Context, projectID string, user mdbv1.AtlasDatabaseUser) workflow.Result {
-	allClustersInProject, _, err := ctx.Client.Clusters.List(context.Background(), projectID, &mongodbatlas.ListOptions{})
+	allClusterNames, err := atlascluster.GetAllClusterNames(ctx.Client, projectID)
```
Comment (chatton) [inline pkg/controller/atlasdatabaseuser/databaseuser.go:164]: we don't actually need the cluster types, we only care about their names

File: test/e2e/actions/actions.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ func UpdateCluster(newData *model.TestDataProvider) {
 	})
 
 	By("Check attributes\n", func() {
-		uCluster := mongocli.GetClustersInfo(newData.Resources.ProjectID, newData.Resources.Clusters[0].Spec.ClusterSpec.Name)
+		uCluster := mongocli.GetClustersInfo(newData.Resources.ProjectID, newData.Resources.Clusters[0].Spec.GetClusterName())
```
Comment (chatton) [inline test/e2e/actions/actions.go:37]: in all places we use ClusterSpec.Name, we use the new access method which checks for Advanced/Regular names

File: test/e2e/actions/steps.go [hunk 4]
```diff
@@ -42,9 +45,18 @@ func WaitCluster(input model.UserInputs, generation string) {
 		input.Namespace, input.Clusters[0].GetClusterNameResource()),
 	).Should(Equal("IDLE"), "Kubernetes resource: Cluster status should be IDLE")
 
-	ExpectWithOffset(
-		1, mongocli.GetClusterStateName(input.ProjectID, input.Clusters[0].Spec.ClusterSpec.Name),
-	).Should(Equal("IDLE"), "Atlas: Cluster status should be IDLE")
+	cluster := input.Clusters[0]
+	if cluster.Spec.AdvancedClusterSpec != nil {
```
Comment (chatton) [inline test/e2e/actions/steps.go:49]: the same function handles the special case of advanced clusters

File: test/e2e/actions/steps.go [hunk 5]
```diff
@@ -130,6 +142,61 @@ func CompareClustersSpec(requested model.ClusterSpec, created mongodbatlas.Clust
...
+			ExpectWithOffset(1, specsAreEqual(created.ReplicationSpecs[i].RegionConfigs[key].ElectableSpecs, region.ElectableSpecs)).To(BeTrue(), "Replica Spec: ElectableSpecs is not the same")
+			ExpectWithOffset(1, created.ReplicationSpecs[i].RegionConfigs[key].ProviderName).Should(Equal(region.ProviderName), "Replica Spec: ProviderName is not the same")
+			ExpectWithOffset(1, created.ReplicationSpecs[i].RegionConfigs[key].RegionName).Should(Equal(region.RegionName), "Replica Spec: RegionName is not the same")
+			ExpectWithOffset(1, created.ReplicationSpecs[i].RegionConfigs[key].Priority).Should(Equal(region.Priority), "Replica Spec: Priority is not the same")
+		}
+	}
+}
+
+// specsAreEqual returns true if an atlas spec matches an operator spec.
+func specsAreEqual(atlasSpecs *mongodbatlas.Specs, specs *v1.Specs) bool {
```
Comment (chatton) [inline test/e2e/actions/steps.go:172]: specs are populated with defaults even if not specified.

File: test/e2e/api/atlas/atlas.go [hunk 6]
```diff
@@ -71,5 +75,22 @@ func (a *Atlas) GetPrivateEndpoint(projectID, provider string) ([]mongodbatlas.P
...
+	ginkgoPrettyPrintf(enpointsList, "listing private endpoints in project %s", projectID)
 	return enpointsList, nil
 }
+
+func (a *Atlas) GetAdvancedCluster(projectId, clusterName string) (*mongodbatlas.AdvancedCluster, error) {
+	advancedCluster, _, err := a.Client.AdvancedClusters.Get(context.Background(), projectId, clusterName)
+	if err != nil {
+		return nil, err
+	}
+	ginkgoPrettyPrintf(advancedCluster, "getting advanced cluster %s in project %s", clusterName, projectId)
```
Comment (chatton) [inline test/e2e/api/atlas/atlas.go:87]: this function simulates what we are currently doing when displaying the output of mongocli when we make requests.

File: test/e2e/data/atlascluster_advanced_helm.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+apiVersion: atlas.mongodb.com/v1
```
Comment (chatton) [inline test/e2e/data/atlascluster_advanced_helm.yaml:1]: standard advanced cluster

File: test/e2e/data/atlascluster_advanced_multi_region_helm.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
+apiVersion: atlas.mongodb.com/v1
```
Comment (chatton) [inline test/e2e/data/atlascluster_advanced_multi_region_helm.yaml:1]: advanced cluster spread across different zones

File: test/e2e/actions/steps.go [hunk 9]
```diff
@@ -130,6 +139,28 @@ func CompareClustersSpec(requested model.ClusterSpec, created mongodbatlas.Clust
 	}
 }
 
+func CompareAdvancedClustersSpec(requested model.ClusterSpec, created mongodbatlas.AdvancedCluster) {
+	advancedSpec := requested.AdvancedClusterSpec
+	Expect(created.MongoDBVersion).ToNot(BeEmpty())
+	Expect(created.MongoDBVersion).ToNot(BeEmpty())
```
Comment (leo-ri) [inline test/e2e/actions/steps.go:145]: typo (lines 144-145 is equal)

File: scripts/create_kind_cluster.sh [hunk 10]
```diff
@@ -90,6 +90,14 @@ nodes:
     hostPort: 30013
     listenAddress: "0.0.0.0"
     protocol: TCP
+  - containerPort: 30014
```
Comment (leo-ri) [inline scripts/create_kind_cluster.sh:93]: need to add test/e2e/config/kind.yaml too

**Review by leo-ri — APPROVED** [review]

File: pkg/controller/atlasdatabaseuser/databaseuser.go (hunk 1 above)
Comment (fabritsius) [reply pkg/controller/atlasdatabaseuser/databaseuser.go:155]: If we check errors thought both API version shouldn't we return an error if either of them is an error, not both?
Also I thought that we can get a regular cluster through the advanced API, I can check this.

**Pull request merged**